src/Purdle/UI/Keyboard.purs

```diff
@@ -68,13 +109,27 @@ keyboard = H.mkComponent
                                 <> keyClass k <> " "
                                 <> fromMaybe "key-grey" (colorClass cmap k)
                   , HP.type_ HP.ButtonButton
-                  , HE.onClick \_ -> k
+                  , HE.onClick \_ -> HandleKey k
                   ]
                   [HH.text (showKey k)]
     , eval: H.mkEval $ H.defaultEval
-        { handleAction = H.raise
+        { handleAction = case _ of
+              Init -> do
+                let
+                  fromKeyboardEvent e = case KE.key e of
+                    "Enter" -> Just EnterKey
+                    "Backspace" -> Just BackspaceKey
+                    letter -> LetterKey <$> readLetter letter
+                document <- H.liftEffect $ Window.document =<< HTML.window
+                void $ H.subscribe $
+                  HQE.eventListener
+                    KET.keyup
+                    (HTMLDocument.toEventTarget document)
+                    (map HandleKey <=< map fromKeyboardEvent <<< KE.fromEvent)
+              HandleKey key -> H.raise key
         , handleQuery = case _ of
             KQSetColors cmap r -> Just r <$ put cmap
+        , initialize = Just Init
```

Comment:
Seems like probably want to remove the event listener when this Component is removed from DOM (even though it doesn't seem to ever unmount currently) but I'm not sure of a good way of doing that aside from threading the `SubscriptionId` returned from `H.subscribe` through the action handlers. I'm not experienced with Halogen, so maybe there's a better way.

Reply:
i ended up adding this to the state and adding it to the `finalize` handler in halogen components :)  i think i have a future idea to support modes of play so this might become relevant in that case.